### ISSUE TITLE
Comma decimal separator

### ISSUE THREAD

**Issue by torfason** [vincentarelbundock/modelsummary#145]
I'm running into some issues with using comma as a decimal separator with `msummary()`. Because `msummary()` converts coefficients and other values to numbers internally (using `sprintf()`presumably), one is dependent on it's conversion routine.  

The underlying issue is that locale support is rather poor in both R and in `sprintf()`. A workaround for `sprintf()` of setting `LC_NUMERIC` is suggested [here](https://stackoverflow.com/questions/23586741/change-decimal-character-in-sprintf). There are two things that make this workaround less than ideal:
  
* Setting `LC_NUMERIC` causes an ominous warning that `setting 'LC_NUMERIC' may cause R to function strangely` (see above on locale support in R generally).
* Setting `LC_NUMERIC` does in fact cause changes in the output. See example below:

```r
library(modelsummary)
m <- lm(dist ~ speed, data=cars)

Sys.setlocale("LC_NUMERIC","C")
msummary(m, output="flextable", gof_omit=".")

Sys.setlocale("LC_NUMERIC","es_ES")
msummary(m, output="flextable", gof_omit=".")
```

<img width="510" alt="Screenshot 2020-10-06 at 12 24 19" src="https://user-images.githubusercontent.com/3247510/95201074-edd8d500-07ce-11eb-9fa4-1c2d1f57c158.png">

Perhaps I'm missing another way, and I'm not suggesting the fix is straightforward. The way that we in comma-land unfortunately must work is to do 95% of our work with the standard "." for decimal settings (and other internationalization settings), but then switch only for the output, so using custom locales is seldom a good solution for us. Which is why some way to specify decimal output in numbers is a feature we are very grateful for in output packages.

**Comment by vincentarelbundock** [comment]
Thanks for the report. This is a very interesting problem, but I must admit that I am far from a locale guru.

One possible solution would be to allow users to feed a *function* to the `fmt` argument (in addition to the current string). That way, you could customize the output to your heart's content. For example, you could use the `comma_format` function from the `scales` package to get something like `1,000,001`.

What do you think? (This would be a fair amount of work, but I'm happy to implement this if there is sufficient enthusiasm for the solution.)

**Comment by vincentarelbundock** [comment]
Note that the change in output seems limited to `flextable`. I would consider opening an issue with them.

The warning I don't mind as much, because you can always flip it back after the table is printed and wrap it in:

```r
invisible(suppressWarnings(Sys.setlocale("LC_NUMERIC", "es_ES")))
```

**Comment by torfason** [comment]
Thanks for the quick response. Passing a custom formatting function would be very powerful, but perhaps a bit difficult to use for non-experts. 

An alternative, perhaps simpler to both implement and use, might be to include a parameter for the locale to use for the output (my current workaround is to encapsulate `msummary()` in my own function, so I can do do that without any changes to `modelsummary`). That does however depend on the output table packages (such as `flextable`) not choking on alternative locales.

Probably best to let this one sit for a while, perhaps a locale expert will chime in, or a good solution emerges. In the meantime, it seems like a good idea to report the formatting issues to `flextable`, since it seems like that should not happen in any case. I'll take a look at that.

**Comment by vincentarelbundock** [comment]
Got it. Thanks for sharing your thoughts on this. 

I'm a bit uncomfortable with asking `modelsummary` to change the global locale, but you're right, hopefully some expert will chime in at some point.

Agreed that it would be good to report to flextable.

**Comment by torfason** [comment]
> I'm a bit uncomfortable with asking modelsummary to change the global locale ...

Agreed, it would need to be reliably self-contained (in my encapsulation function I save the prior locale and then switch back afterwards – it seems like it should be safe given that there is only one thread that could be messing with it).

**Comment by torfason** [comment]
tl;dr: I believe after looking at this more closely that the most idiomatic way to address this would be if `modelsummary` could use `getOption("OutDec")` to determine the decimal separator to use in tables.

---

I haver reported an issue for flextable, it's been accepted as a bug (with the caveat that they would like to avoid changes to the locale). In the meantime I've looked more at the issue. Key takeaways:

Setting LC_NUMERIC is not recommended. From the help for `Sys.setlocale()`:

>Warning
>Setting "LC_NUMERIC" to any value other than "C" may cause R to function anomalously, so gives a warning. Input conversions in R itself are unaffected, but the reading and writing of ASCII save files will be, as may packages which do their own input/output.
>
>Setting it temporarily on a Unix-alike to produce graphical or text output may work well enough, but _**options(OutDec) is often preferable**_.
>
>Almost all the output routines used by R itself under Windows ignore the setting of "LC_NUMERIC" since they make use of the Trio library which is not internationalized.

Regarding OutDec, the following comes from the help for `options()`:

>OutDec:
>character string containing a single character. _**The preferred character to be used as the decimal point in output conversions**_, that is in printing, plotting, format and as.character but not when deparsing nor by sprintf nor formatC (which are sometimes used prior to printing.)

It turns out that most of the output table packages targeted by modelsummary do respect `OutDec`. Reproducible demo below (but output not included to save space):

```
d <- head(trees, 3)
options(OutDec = ",")

#' # Respect options(OutDec)
knitr::kable(d)
flextable::flextable(d)
huxtable::hux(d)
gt::gt(d)
kableExtra::kable(d)

#' # Do not respect options(OutDec)
htmlTable::htmlTable(d)
tibble::tibble(d)
```

Due to the required formatting, `modelsummary` converts numbers to character before passing values to target packages, so it does not benefit from the target packages using `getOption("OutDec")`. Therefore, I think the most idiomatic approach to resolve this issue would be if `modelsummary` could respect this option as well. Unfortunately, `sprintf` _does not_, so this would probably require postprocessing after `sprintf` does its thing.

**Comment by vincentarelbundock** [comment]
Thanks a lot for digging into this @torfason , I really appreciate it!

What I take out of this discussion is that `sprintf` is not a good default, because it is not flexible enough. I have just pushed a new version to Github which adds a lot of flexibility to the `fmt` argument. Changes:

1. Default value: `fmt=3`. This determines the number of digits after the dot, keeping zeros. It also respects `options(OutDec)` by calling `format` instead of `sprintf`.
2. `fmt` also accepts a character string, which is passed to `sprintf`, so legacy code will still be supported.
3. It can also take a custom function.

I would really appreciate it if you could install the github version, look at the new documentation (`?modelsummary`) and let me know what you think. To install the development version of `modelsummary`:

```r
library(remotes)
install_github("vincentarelbundock/modelsummary")
```

Then restart your R session completely to make sure that the changes take effect.

This should allow you to do this:

    library(modelsummary)
    mod <- lm(mpg ~ hp, mtcars)

    options(OutDec=",")

    # integer
    modelsummary(mod, fmt=2, output="markdown")

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Model 1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: left;">30,10</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(1,63)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">hp</td>
<td style="text-align: left;">-0,07</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0,01)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Num.Obs.</td>
<td style="text-align: left;">32</td>
</tr>
<tr class="even">
<td style="text-align: left;">R2</td>
<td style="text-align: left;">0,602</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R2 Adj.</td>
<td style="text-align: left;">0,589</td>
</tr>
<tr class="even">
<td style="text-align: left;">AIC</td>
<td style="text-align: left;">181,2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BIC</td>
<td style="text-align: left;">185,6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Log.Lik.</td>
<td style="text-align: left;">-87,619</td>
</tr>
<tr class="odd">
<td style="text-align: left;">F</td>
<td style="text-align: left;">45,460</td>
</tr>
</tbody>
</table>


    # string
    modelsummary(mod, fmt="%.5f", output="markdown")

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Model 1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: left;">30.09886</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(1.63392)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">hp</td>
<td style="text-align: left;">-0.06823</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0.01012)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Num.Obs.</td>
<td style="text-align: left;">32</td>
</tr>
<tr class="even">
<td style="text-align: left;">R2</td>
<td style="text-align: left;">0,602</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R2 Adj.</td>
<td style="text-align: left;">0,589</td>
</tr>
<tr class="even">
<td style="text-align: left;">AIC</td>
<td style="text-align: left;">181,2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BIC</td>
<td style="text-align: left;">185,6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Log.Lik.</td>
<td style="text-align: left;">-87,619</td>
</tr>
<tr class="odd">
<td style="text-align: left;">F</td>
<td style="text-align: left;">45,460</td>
</tr>
</tbody>
</table>


    # function
    noise <- function(x) round(runif(length(x)), 3)
    modelsummary(mod, fmt=noise, output="markdown")

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Model 1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: left;">0,906</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0,59)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">hp</td>
<td style="text-align: left;">0,933</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0,038)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Num.Obs.</td>
<td style="text-align: left;">32</td>
</tr>
<tr class="even">
<td style="text-align: left;">R2</td>
<td style="text-align: left;">0,602</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R2 Adj.</td>
<td style="text-align: left;">0,589</td>
</tr>
<tr class="even">
<td style="text-align: left;">AIC</td>
<td style="text-align: left;">181,2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">BIC</td>
<td style="text-align: left;">185,6</td>
</tr>
<tr class="even">
<td style="text-align: left;">Log.Lik.</td>
<td style="text-align: left;">-87,619</td>
</tr>
<tr class="odd">
<td style="text-align: left;">F</td>
<td style="text-align: left;">45,460</td>
</tr>
</tbody>
</table>

<sup>Created on 2020-10-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by torfason** [comment]
"fmt is now powerful" indeed! :-)

That was incredibly quick and thorough. Thanks! I've installed the most recent version from GitHub and it looks like it's working flawlessly. I'll be using this version from now on, and if there are any unexpected issues I'll make sure to report back.

As an aside, I was thinking about why I had just come across this issue now, given that I've needed to format tables with commas from R for quite a number of years. The answer turned out to be that until now, I have always had to output model tables to Excel for final formatting anyway, so the commas were a minor annoyance. Now I was able to produce an end result that needed no manual formatting apart from the commas, making them more salient. And with this new version of modelsummary, it looks like I am finally all the way there. Thanks again!

**Comment by vincentarelbundock** [comment]
This is excellent news!

**Comment by vincentarelbundock** [comment]
By the way, I was just about to send my book (in French) to press, and all my code examples still had dots instead of commas. Now that I learned about `options(OutDec)`, my code examples won't be so horrible anymore ;)

Thanks!
